jupyter_server/auth/identity.py

```diff
@@ -0,0 +1,155 @@
+
+    - :meth:`~.IdentityProvider.get_user` returns a user object.
+      For successful authentication,
+      this may return anything truthy.
+
+    - :meth:`~.IdentityProvider.identity_model` returns a standard identity model dictionary,
+      for use in the /me API.
+      This should accept whatever is returned from get_user()
+      and return a dictionary matching the structure of
+      :class:`~.IdentityModel`.
+
+    .. versionadded:: 2.0
+    """
+
+    def get_user(self, handler: RequestHandler) -> Any:
+        """Get the authenticated user for a request
+
+        User may be anything truthy, but must be understood by identity_model method.
+
+        Return None if the request is not authenticated.
+
+        When in doubt, use a standard identity model.
+        """
+
+        if handler.login_handler is None:
+            return {
+                "username": "anonymous",
+            }
+
+        # The default: call LoginHandler.get_user for backward-compatibility
+        # TODO: move default implementation to this class,
+        # deprecate `LoginHandler.get_user`
+        user = handler.login_handler.get_user(handler)
+        return user
+
+    def identity_model(self, user: Any) -> IdentityModel:
+        """Construct standardized identity model for the identity API
+
+        Casts objects returned by `.get_user` (generally str username or dict with 'username' or 'name')
+        To a complete IdentityModel dict.
+
+        `username` is required.
+        Any other missing fields will be filled out with defaults.
+        """
+        identity = {}
+        if isinstance(user, str):
+            return {
+                "username": user,
+            }
+        elif isinstance(user, dict):
+            # username may be in 'username' field or 'name' (e.g. JupyterHub)
+            # but only accept 'name' for username if 'username' not present
+            for username_key in ("username", "name"):
+                if username_key in user:
+                    identity["username"] = user[username_key]
+                    break
+
+            for key, value in user.items():
+                # annotations is where fields are stored
+                if key in IdentityModel.__annotations__:
+                    identity[key] = user[key]
+
+        # handle other types, e.g. custom objects. Subclasses must define this method
+        # in order to handler these.
+        if "username" not in identity:
+            clsname = self.__class__.__name__
+            self.log.warning(
+                f"Unable to find username in current_user. {clsname}.identity_model() must accept user objects as returned by {clsname}.get_user()."
+            )
+            self.log.debug("Unable to find username in current_user: %s", user)
+            identity["username"] = "unknown"
+        # fill defaults
+        return self.fill_identity(identity)
+
+    def _get_identity_model(self, user: Any) -> IdentityModel:
```

Comment:
I was thinking about subclasses - having this method means subclasses must provide a _fillable_ model, wheeras if we always call `get_identity` directly, they must provide a _filled_ model. That's a backward-compatiblity problem if we ever define new fields that can be filled with default values.

I do think it's important that `jupyter-server` call fill_identity on whatever is the recommended "implementer's should override this" method. But maybe there's a better way to organize/name those two methods, e.g. more straightforward and descriptive `get_identity` and `get_filled_identity`.

Reply:
If consumers should always get a filled model, but suppliers should not need to care about that, it might make sense to make the (overridable) supplier be `_get_identity` in order to signal that it is not meant for consumers ?

Reply:
I switched it to a User dataclass, which means I can put `fill_defaults` in `__post_init__`, removing this issue, as the User class solves:

1. get_user always returns a filled object from e.g. `User(username="abc")`, and
2. IdentityProvider subclasses don't need to implement populating defaults (though they _may_ in User subclasses)